Document release process

lydia-duncan opened the pull request.

lydia-duncan committed:
Add steps for how to release to the README

----
Signed-off-by: Lydia Duncan <lydia-duncan@users.noreply.github.com>

lydia-duncan committed:
Fix the bullets to use the proper format

Had to look it up

----
Signed-off-by: Lydia Duncan <lydia-duncan@users.noreply.github.com>

jabraham17 commented on README.rst:
```diff
@@ -31,6 +31,31 @@ To install from source on github_::
+Release
+-------
+
+#. Test and commit changes
+#. Merge your changes
```
This should not be a part of the release process

lydia-duncan replied:
I've been considering it part of the release process since changes frequently should go live right away, given that there's not a regular release cadence for this repository

jabraham17 replied:
sure, that has been the status quo. But its not necessarily true and it seems misleading to start a section about "release process" with "make the change"

lydia-duncan replied:
Do you see it changing in the future?

I'd be willing to add some wording about that expectation, maybe something about "changes or sets of changes in a short period of time", but I think this documentation should reflect the most reasonable expected use case and my expectation is that releases would be motivated by single to grouped changes rather than an explicit cadence for the foreseeable future.

jabraham17 replied:
I can absolutely see myself batching a set of changes before pushing them out as a relase

lydia-duncan replied:
Clarifying: you're aware I'm not trying to discourage that, right?  What I'm trying to discourage is someone making a change and holding off on making a release until another change goes live, but then failing to get the other change in, leaving the first change unavailable for potentially a year or more, at which point maybe it doesn't work anymore since it wasn't incorporated into the Chapel repository (the main place this repo is used)

lydia-duncan replied:
We have had a period in the past where the Chapel repository relied on a checkout of this repo instead of a release, but it is generally better for changes to be out in a release.  This is because:
- There may be projects relying on it that we're not aware of, and such projects should benefit from changes made to benefit chpldoc
- We want to make sure the release process still works.  While having that be discovered as a result of other unrelated changes is not great, it is still better than discovering it doesn't work years after the fact
  - This is because it may be harder to find *why* a failure happened if a lot of changes have happened in the interim
  - And when a change is needed due to a deprecation, those nice deprecation messages do not last forever.  The closer to when a deprecation happens, the more likely it will be that tools to aid in that transition still work, making it easier to make any needed updates

jabraham17 replied:
>  What I'm trying to discourage is someone making a change and holding off on making a release until another change goes live, but then failing to get the other change in, leaving the first change unavailable for potentially a year or more, at which point maybe it doesn't work anymore since it wasn't incorporated into the Chapel repository (the main place this repo is used)

I understand that concern. My concern is the wording here

```
Release
-------

#. Test and commit changes
#. Merge your changes
```

This to me implies that development and release are the same thing, which is not true.


Consider the following structure instead

```
Making Changes
-----------------

#. Test and commit changes
#. Merge your changes
#. Generate a new PyPi release

Making a Release
-----------------
#. Go to main page for repo
```

This clearly delineates between development and the process of making a release, without tying the two together (and hopefully addressing your concern)

jabraham17 commented on README.rst:
```diff
@@ -31,6 +31,31 @@ To install from source on github_::
+   (otherwise we’ll have build issues), so it should be straight-forward to
+   figure out how to do this
+
+IF SOMETHING GOES WRONG
+
+The release pushing job is in .github/workflows/python-publish.yml
```
Rather than the all caps "if something goes wrong", consider a note like "the Github Action for publishing a new version lives in .github/workflows/python-publish.yml"

lydia-duncan committed:
Add advice on how to fix problems with the release

----
Signed-off-by: Lydia Duncan <lydia-duncan@users.noreply.github.com>

lydia-duncan committed:
Modifications based on Jade's review

- Separate the release process into a section for making changes and a section
  for making a release, with an indication in the former that we expect the
  latter to follow quickly after
- reformat the portion about how to handle issues with the release script

----
Signed-off-by: Lydia Duncan <lydia-duncan@users.noreply.github.com>

lydia-duncan committed:
Formatting adjustment to the note about if things go wrong

`.. note` is a Sphinx directive, not a .rst one.

----
Signed-off-by: Lydia Duncan <lydia-duncan@users.noreply.github.com>
